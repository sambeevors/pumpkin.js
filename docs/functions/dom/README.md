---
next: /functions/string/
prev: /functions/array/
---

# DOM functions

The following functions concern DOM manipulation including events.

## qs

Get an element in the DOM

### Parameters

- `selector` (type: String) - The selector to match against
- `scope` (type: Node) - An element to search within

### Returns

- (type: Node) - The first matching element

### Example

```js
$foo = $.qs('.foo')
```

## qsa

Get all matching elements in the DOM as an array

### Parameters

- `selector` (type: String) - The selector to match against
- `scope` (type: Node) - An element to search within

### Returns

- (type: Array) - The matching elements

:::tip
Because `qsa` returns as an array, you have access to [array prototype methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/prototype)
:::

### Example

```js
$foo = $.qsa('.foo')
```

## children

Get the children of an element

### Parameters

- `el` (type: Node) - The element to search

### Returns

- (type: Array) - The children of the passed element

### Example

```js
$foo = $.qs('.foo')
$fooChildren = $.children($foo)
```

## siblings

Get the siblings of an element

### Parameters

- `el` (type: Node) - The element to search

### Returns

- (type: Array) - The siblings of the passed element

### Example

```js
$foo = $.qs('.foo')
$fooSiblings = $.siblings($foo)
```

## contains

Check if element contains child

### Parameters

- `el` (type: Node) - The element to search
- `child` (type: Node) - The child element to search for

### Returns

- (type: Boolean) - Whether or not child was found

### Example

```js
$foo = $.qs('.foo')
$bar = $.qs('.bar')

console.log($.contains($foo, $bar))
```

## empty

Removes all children from element

### Parameters

- `el` (type: Node) - The element to empty

### Example

```js
$foo = $.qs('.foo')
$.empty($foo)
```

## remove

Remove element from the DOM

### Parameters

- `el` (type: Node) - The element to remove

### Example

```js
$foo = $.qs('.foo')
$.remove($foo)
```

## fromTop

Get the distance between an element and the top of the page

### Parameters

- `el` (type: Node) - The element to check

### Returns

- (type: Number) - The distance from the top of the page (in pixels)

### Example

```js
$foo = $.qs('.foo')
$.fromTop($foo) // 100
```

## show

Show selected element

### Parameters

- `el` (type: Node) - The element to show

### Example

```js
$foo = $.qs('.foo')
$.show($foo)
```

## hide

Hide selected element

### Parameters

- `el` (type: Node) - The element to hide

### Example

```js
$foo = $.qs('.foo')
$.hide($foo)
```

## toggle

Toggle visibility of selected element

### Parameters

- `el` (type: Node) - The element to toggle

### Example

```js
$foo = $.qs('.foo')
$.toggle($foo)
```

## css

Apply multiple inline styles to an element

### Parameters

- `el` (type: Node) - The element to apply the styles to
- `obj` (type: Object) - An object of CSS rules

### Example

```js
$foo = $.qs('.foo')
$.css($foo, {
  backgroundColor: 'red',
  fontWeight: 'bold'
})
```

## on

Apply propagating event handlers

### Parameters

- `event` (type: String) - The event to trigger on
- `array` (type: Array) - The array of nodes to apply to
- `selector` (type: String) - Query selector for event to bind to [optional]
- `cb` (type: Function) - The function to run upon event

### Example

```js
$foo = $.qsa('.foo')

$.on('click', $foo, 'a', e => {
  console.log(e)
})
```

## ready

Run callback function once DOM has fully loaded

### Parameters

- `cb` (type: Function) - The function to run once ready

### Example

```js
$.ready(() => {
  console.log('DOM is ready')
})
```
