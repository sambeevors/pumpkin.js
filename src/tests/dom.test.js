import * as $ from '../dom'
import DOM from './helpers/fake-dom'

let $container
let $toDoList
let $toDoListItems
let $toDoListItem
let $hiddenElement

const dom = new DOM(
  `<div id="container">
    <h1>To Do</h1>
    <ul id="to-do-list">
      <li>Go shopping</li>
      <li>Do laundry</li>
      <li>Raid fridge!</li>
    </ul>
    <div id="hidden-element" style="display: none;"></div>
  </div>`,
  () => {
    $container = $.qs('#container')
    $toDoList = $.qs('#to-do-list')
    $toDoListItems = $.qsa('li', $toDoList)
    $toDoListItem = $.qs('li', $toDoList)
    $hiddenElement = $.qs('#hidden-element')
  }
)

test('Query selector', () => {
  expect($toDoList).toMatchSnapshot()
})

test('Query selector all', () => {
  expect($toDoListItems).toMatchSnapshot()
})

test('Get child nodes of an element', () => {
  // Basic test
  expect($.children($toDoList)).toMatchSnapshot()
  // Test element with no children
  expect($.children($hiddenElement)).toHaveLength(0)
})

test('Get sibling nodes of an element', () => {
  // Basic test
  expect($.siblings($toDoListItem)).toMatchSnapshot()
  // Test element with no siblings
  expect($.siblings($container)).toHaveLength(0)
})

test('Check if node contains specific child node', () => {
  // Check where true
  expect($.contains($toDoList, $toDoListItem)).toBe(true)
  // Check where false
  expect($.contains($toDoList, $hiddenElement)).toBe(false)
})

test('Empty element of all children', () => {
  $.empty($toDoList)
  expect($.children($toDoList)).toHaveLength(0)

  dom.resetDom()
})

test('Remove element from DOM', () => {
  $.remove($toDoList)
  expect($.qs('#to-do-list')).toBeFalsy()

  dom.resetDom()
})

test('Get element distance from top of the page', () => {
  expect($.fromTop($toDoList)).toBeDefined()
})

test('Visually show an element', () => {
  $.show($hiddenElement)
  expect($hiddenElement.style.display).toBe('')

  dom.resetDom()
})

test('Visually hide an element', () => {
  $.hide($toDoList)
  expect($toDoList.style.display).toBe('none')

  dom.resetDom()
})

test('Toggle element visibility', () => {
  $.toggle($hiddenElement)
  expect($hiddenElement.style.display).toBe('')

  $.toggle($hiddenElement)
  expect($hiddenElement.style.display).toBe('none')

  dom.resetDom()
})

test('Add CSS to an element', () => {
  $.css($toDoList, {
    margin: '20px auto',
    fakeProperty: 'foo'
  })
  expect($toDoList).toMatchSnapshot()

  dom.resetDom()
})

test('Initiate a propagating event listener on an element', () => {
  // Setup click event
  let clickEvent = new MouseEvent('click', {
    view: window,
    bubbles: true,
    cancelable: true
  })
  let triggered = false

  // Basic test
  $.on('click', $toDoList, () => {
    triggered = true
  })
  $toDoList.dispatchEvent(clickEvent)
  expect(triggered).toBeTruthy()

  triggered = false

  // Test with propagation
  $.on('click', $toDoList, 'li', () => {
    triggered = true
  })
  $toDoListItem.dispatchEvent(clickEvent)
  expect(triggered).toBeTruthy()

  triggered = false

  // Test clicking wrong item
  $.on('click', $container, 'h1', () => {
    triggered = true
  })
  $hiddenElement.dispatchEvent(clickEvent)
  expect(triggered).toBeFalsy()
})

test('Check when DOM is ready', () => {
  let ready = false
  $.ready(() => {
    ready = true
  })
  expect(ready).toBeTruthy()
})
