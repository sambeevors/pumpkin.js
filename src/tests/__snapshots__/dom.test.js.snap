// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add CSS to an element 1`] = `
<ul
  id="to-do-list"
  style="margin: 20px auto;"
>
  
      
  <li>
    Go shopping
  </li>
  
      
  <li>
    Do laundry
  </li>
  
      
  <li>
    Raid fridge!
  </li>
  
    
</ul>
`;

exports[`Get child nodes of an element 1`] = `
Array [
  <li>
    Go shopping
  </li>,
  <li>
    Do laundry
  </li>,
  <li>
    Raid fridge!
  </li>,
]
`;

exports[`Get sibling nodes of an element 1`] = `
Array [
  <li>
    Do laundry
  </li>,
  <li>
    Raid fridge!
  </li>,
]
`;

exports[`Query selector 1`] = `
<ul
  id="to-do-list"
>
  
      
  <li>
    Go shopping
  </li>
  
      
  <li>
    Do laundry
  </li>
  
      
  <li>
    Raid fridge!
  </li>
  
    
</ul>
`;

exports[`Query selector all 1`] = `
Array [
  <li>
    Go shopping
  </li>,
  <li>
    Do laundry
  </li>,
  <li>
    Raid fridge!
  </li>,
]
`;
